File: examples/leetcode-out/php/10/regular-expression-matching.php
<?php
function mochi_isMatch($s, $p) {
	$m = (is_array($s) ? count($s) : strlen($s));
	$n = (is_array($p) ? count($p) : strlen($p));
	$dp = [];
	$i = 0;
	while (($i <= $m)) {
		$row = [];
		$j = 0;
		while (($j <= $n)) {
			$row = ((is_array($row) && is_array([false])) ? array_merge($row, [false]) : ((is_string($row) || is_string([false])) ? ($row . [false]) : ($row + [false])));
			$j = ((is_array($j) && is_array(1)) ? array_merge($j, 1) : ((is_string($j) || is_string(1)) ? ($j . 1) : ($j + 1)));
		}
		$dp = ((is_array($dp) && is_array([$row])) ? array_merge($dp, [$row]) : ((is_string($dp) || is_string([$row])) ? ($dp . [$row]) : ($dp + [$row])));
		$i = ((is_array($i) && is_array(1)) ? array_merge($i, 1) : ((is_string($i) || is_string(1)) ? ($i . 1) : ($i + 1)));
	}
	$dp[$m][$n] = true;
	$i2 = $m;
	while (($i2 >= 0)) {
		$j2 = ($n - 1);
		while (($j2 >= 0)) {
			$first = false;
			if (($i2 < $m)) {
				if (((($p[$j2] == $s[$i2])) || (($p[$j2] == ".")))) {
					$first = true;
				}
			}
			$star = false;
			if ((((is_array($j2) && is_array(1)) ? array_merge($j2, 1) : ((is_string($j2) || is_string(1)) ? ($j2 . 1) : ($j2 + 1))) < $n)) {
				if (($p[((is_array($j2) && is_array(1)) ? array_merge($j2, 1) : ((is_string($j2) || is_string(1)) ? ($j2 . 1) : ($j2 + 1)))] == "*")) {
					$star = true;
				}
			}
			if ($star) {
				$ok = false;
				if ($dp[$i2][((is_array($j2) && is_array(2)) ? array_merge($j2, 2) : ((is_string($j2) || is_string(2)) ? ($j2 . 2) : ($j2 + 2)))]) {
					$ok = true;
				} else {
					if ($first) {
						if ($dp[((is_array($i2) && is_array(1)) ? array_merge($i2, 1) : ((is_string($i2) || is_string(1)) ? ($i2 . 1) : ($i2 + 1)))][$j2]) {
							$ok = true;
						}
					}
				}
				$dp[$i2][$j2] = $ok;
			} else {
				$ok = false;
				if ($first) {
					if ($dp[((is_array($i2) && is_array(1)) ? array_merge($i2, 1) : ((is_string($i2) || is_string(1)) ? ($i2 . 1) : ($i2 + 1)))][((is_array($j2) && is_array(1)) ? array_merge($j2, 1) : ((is_string($j2) || is_string(1)) ? ($j2 . 1) : ($j2 + 1)))]) {
						$ok = true;
					}
				}
				$dp[$i2][$j2] = $ok;
			}
			$j2 = ($j2 - 1);
		}
		$i2 = ($i2 - 1);
	}
	return $dp[0][0];
}

function mochi_test_example_1() {
	if (!((mochi_isMatch("aa", "a") == false))) { throw new Exception('expect failed'); }
}

function mochi_test_example_2() {
	if (!((mochi_isMatch("aa", "a*") == true))) { throw new Exception('expect failed'); }
}

function mochi_test_example_3() {
	if (!((mochi_isMatch("ab", ".*") == true))) { throw new Exception('expect failed'); }
}

function mochi_test_example_4() {
	if (!((mochi_isMatch("aab", "c*a*b") == true))) { throw new Exception('expect failed'); }
}

function mochi_test_example_5() {
	if (!((mochi_isMatch("mississippi", "mis*is*p*.") == false))) { throw new Exception('expect failed'); }
}

mochi_test_example_1();
mochi_test_example_2();
mochi_test_example_3();
mochi_test_example_4();
mochi_test_example_5();

File: examples/leetcode-out/pas/10/regular-expression-matching.pas
program main;
{$mode objfpc}

uses SysUtils, fgl, fphttpclient, Classes, Variants, fpjson, jsonparser;

type 
  generic TArray<T> = array of T;

function isMatch(s: string; p: string): boolean;

var 
  dp: specialize TArray<specialize TArray<boolean>>;
  first: boolean;
  i: integer;
  i2: integer;
  j: integer;
  j2: integer;
  m: integer;
  n: integer;
  ok: boolean;
  row: specialize TArray<boolean>;
  star: boolean;
begin
  m := Length(s);
  n := Length(p);
  dp := specialize TArray<specialize TArray<boolean>>([]);
  i := 0;
  while (i <= m) do
    begin
      row := specialize TArray<boolean>([]);
      j := 0;
      while (j <= n) do
        begin
          row := Concat(row, specialize TArray<boolean>([False]));
          j := j + 1;
        end;
      dp := Concat(dp, specialize TArray<integer>([row]));
      i := i + 1;
    end;
  dp[m][n] := True;
  i2 := m;
  while (i2 >= 0) do
    begin
      j2 := n - 1;
      while (j2 >= 0) do
        begin
          first := False;
          if (i2 < m) then
            begin
              if ((_indexString(p, j2) = _indexString(s, i2)) or (_indexString(p, j2) = '.')) then
                begin
                  first := True;
                end;
            end;
          star := False;
          if (j2 + 1 < n) then
            begin
              if (_indexString(p, j2 + 1) = '*') then
                begin
                  star := True;
                end;
            end;
          if star then
            begin
              ok := False;
              if dp[i2][j2 + 2] then
                begin
                  ok := True;
                end
              else
                begin
                  if first then
                    begin
                      if dp[i2 + 1][j2] then
                        begin
                          ok := True;
                        end;
                    end;
                end;
              dp[i2][j2] := ok;
            end
          else
            begin
              ok := False;
              if first then
                begin
                  if dp[i2 + 1][j2 + 1] then
                    begin
                      ok := True;
                    end;
                end;
              dp[i2][j2] := ok;
            end;
          j2 := j2 - 1;
        end;
      i2 := i2 - 1;
    end;
  result := dp[0][0];
  exit;
end;

procedure test_example_1;
begin
  if not ((isMatch('aa', 'a') = False)) then raise Exception.Create('expect failed');
end;

procedure test_example_2;
begin
  if not ((isMatch('aa', 'a*') = True)) then raise Exception.Create('expect failed');
end;

procedure test_example_3;
begin
  if not ((isMatch('ab', '.*') = True)) then raise Exception.Create('expect failed');
end;

procedure test_example_4;
begin
  if not ((isMatch('aab', 'c*a*b') = True)) then raise Exception.Create('expect failed');
end;

procedure test_example_5;
begin
  if not ((isMatch('mississippi', 'mis*is*p*.') = False)) then raise Exception.Create(
                                                                                     'expect failed'
    );
end;

function _indexString(s: string; i: integer): string;
begin
  if i < 0 then i := Length(s) + i;
  if (i < 0) or (i >= Length(s)) then
    raise Exception.Create('index out of range');
  Result := s[i + 1];
end;

begin
  test_example_1;
  test_example_2;
  test_example_3;
  test_example_4;
  test_example_5;
end.
